### ISSUE TITLE
Unnecessary re-hashing of keys in `instance Hashable (HashMap k v)`

### ISSUE THREAD

**Issue by sjakobi** [haskell-unordered-containers/unordered-containers#402]
https://github.com/haskell-unordered-containers/unordered-containers/blob/b6bde46eae679111c3f088964f192290cc08b6c8/Data/HashMap/Internal.hs#L508-L530

For `Leaf` nodes, we skip hashing the keys. For `Collision` nodes we hash the keys. I think we should skip hashing the latter too.

**Comment by sjakobi** [comment]
It might be useful to hash the length of the `Collision` arrays though. For e.g. `ByteString`, these are hashed too.

**Comment by treeowl** [comment]
Good point. Another option is to get a fold that includes the hash and use that. Less efficient for collisions, but we assume those are rare.

**Comment by sjakobi** [comment]
> Another option is to get a fold that includes the hash and use that. Less efficient for collisions, but we assume those are rare.

I don't understand what kind of fold would work here in combination with the necessary sorting of hashes?
